libs/s25main/world/GameWorld.cpp

```diff
@@ -413,43 +413,43 @@ void GameWorld::RecalcTerritory(const noBaseBuilding& building, TerritoryChangeR
             sizeChanges[oldOwner - 1]--;
     }
 
-    std::set<MapPoint, MapPointLess> ptsHandled;
+    const std::vector<MapPoint> ptsToHandle = GetAllNeighboursUnion(ptsWithChangedOwners);
+
     // Destroy everything from old player on all nodes where the owner has changed
-    for(const MapPoint& curMapPt : ptsWithChangedOwners)
+    for(const MapPoint& curMapPt : ptsToHandle)
     {
-        // Destroy everything around this point as this is at best a border node where nothing should be around
         // Do not destroy the triggering building or its flag
-        // TODO: What about this point?
-        const uint8_t owner = GetNode(curMapPt).owner;
-        for(const MapPoint neighbourPt : GetNeighbours(curMapPt))
-        {
-            if(ptsHandled.insert(neighbourPt).second)
-                DestroyPlayerRests(neighbourPt, owner, &building);
-        }
+        DestroyPlayerRests(curMapPt, GetNode(curMapPt).owner, &building);
 
         if(gi)
             gi->GI_UpdateMinimap(curMapPt);
     }
 
-    // Destroy remaining roads going through non-owned territory
-    for(const MapPoint& curMapPt : ptsWithChangedOwners)
+    // Destroy remaining roads going through non-owned and border territory
+    for(const MapPoint& curMapPt : ptsToHandle)
     {
         // Skip if there is an object. We are looking only for roads going through, not ending here
         // (objects here are already destroyed and if the road ended there it would have been as well)
         if(GetNode(curMapPt).obj)
             continue;
         Direction dir;
         noFlag* flag = GetRoadFlag(curMapPt, dir);
-        if(!flag || flag->GetPlayer() + 1 == GetNode(curMapPt).owner)
+
+        if(!flag)
+            continue;
+
+        const uint8_t owner = GetNode(curMapPt).owner;
+        if(flag->GetPlayer() + 1 == owner && !IsBorderNode(curMapPt, owner))
```

Comment:
Even better (and faster):
```suggestion
        if(flag->GetPlayer() + 1 == owner && IsPlayerTerritory(curMapPt, owner))
```

`IsBorderNode = GetNode(pt).owner == owner && !IsPlayerTerritory(pt, owner))`   
--> `!IsBorderNode = GetNode(pt).owner != owner || IsPlayerTerritory(pt, owner))`   and the first part can never be true here